derive_client/utils/w3.py

```diff
@@ -229,6 +237,63 @@ def wait_for_tx_receipt(w3: Web3, tx_hash: str, timeout=120, poll_interval=1) ->
         time.sleep(poll_interval)
 
 
+def wait_for_tx_finality(
+    w3: Web3,
+    tx_hash: str,
+    finality_blocks: int = 10,
+    timeout: float = 300.0,
+    poll_interval: float = 1.0,
+):
+    """
+    Wait until tx is mined and has `finality_blocks` confirmations.
+    On timeout this raises one of:
+      - FinalityTimeout: receipt exists but not enough confirmations
+      - TxPendingTimeout: no receipt, but tx present and pending in mempool
+      - TransactionDropped: no receipt and tx not known to node (likely dropped)
+    """
+
+    start_time = time.monotonic()
+
+    while True:
+        try:
+            receipt = AttributeDict(w3.eth.get_transaction_receipt(tx_hash))
+        # receipt can disappear temporarily during reorgs, or if RPC provider is not synced
+        except Exception:
+            receipt = None
+        # blockNumber can change as tx gets reorged into different blocks
+        if receipt is not None and (block_number := w3.eth.block_number) >= receipt.blockNumber + finality_blocks:
+            return receipt
```

Comment:
the walrus operator offers us a nifty solution:
- we only call `w3.eth.block_number` if we indeed have a receipt
- we assign `block_number`, and log it on timeout when  `if receipt is not None` holds, and hence we know it must have been assigned. I'm sure no type checker would approve :smiling_face_with_tear: